File: app/src/main/java/com/spiderbiggen/randomchampionselector/ui/activities/ChampionActivity.java
package com.spiderbiggen.randomchampionselector.ui.activities;

import android.content.Intent;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.support.v7.app.ActionBar;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.spiderbiggen.randomchampionselector.R;
import com.spiderbiggen.randomchampionselector.ddragon.DDragon;
import com.spiderbiggen.randomchampionselector.model.Champion;
import com.spiderbiggen.randomchampionselector.model.ImageType;
import com.spiderbiggen.randomchampionselector.storage.database.DatabaseManager;

import java.io.IOException;

import io.reactivex.disposables.Disposable;

public class ChampionActivity extends ButtonActivity {

    public static final String CHAMPION_KEY = "CHAMPION_KEY";
    public static final String UP_ON_BACK_KEY = "UP_ON_BACK_KEY";
    private static final String TAG = ChampionActivity.class.getSimpleName();
    private static final ImageType imageType = ImageType.SPLASH;
    private int championKey = -1;
    private boolean upOnBack;
    private Disposable championFlowable;
    private boolean shouldFinalize = false;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        setContentView(R.layout.activity_champion);
        super.onCreate(savedInstanceState);
        supportPostponeEnterTransition();
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setTitle(null);
        }

        if (savedInstanceState != null) {
            championKey = savedInstanceState.getInt(CHAMPION_KEY);
        }
        Intent intent = getIntent();
        championKey = intent.getIntExtra(CHAMPION_KEY, championKey);
        upOnBack = intent.getBooleanExtra(UP_ON_BACK_KEY, true);
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        outState.putInt(CHAMPION_KEY, championKey);
        super.onSaveInstanceState(outState);
    }

    @Override
    protected void onResume() {
        DatabaseManager dbInstance = DatabaseManager.getInstance();
        championFlowable = championKey < 0
                ? dbInstance.findRandomChampion(this::setChampion, null, championKey)
                : dbInstance.findChampion(this::setChampion, championKey);
        super.onResume();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                onBackPressed();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onBackPressed() {
        if (upOnBack) {
            Intent parentActivityIntent = getParentActivityIntent();
            parentActivityIntent = parentActivityIntent == null ? new Intent(this, ListChampionsActivity.class) : parentActivityIntent;
            supportNavigateUpTo(parentActivityIntent);
            overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out);
        } else {
            supportFinishAfterTransition();
        }
    }

    @Override
    protected void onPause() {
        if (championFlowable != null && !championFlowable.isDisposed()) {
            championFlowable.dispose();
        }
        super.onPause();
    }

    @Override
    protected void onStop() {
        super.onStop();
        if (shouldFinalize) {
            finish();
        }
    }

    @Override
    public void openChampion(View view) {
        super.openChampion(view);
        shouldFinalize = true;
    }

    public void setChampion(Champion champion) {
        if (champion == null) {
            championKey = -1;
            return;
        }
        championKey = champion.getKey();
        Bitmap bitmap = null;
        try {
            bitmap = DDragon.getInstance().getChampionBitmap(champion, imageType);
        } catch (IOException e) {
            Log.e(TAG, "setChampion: ", e);
        }
        ImageView bg = findViewById(R.id.champion_splash);
        if (bg != null && bitmap != null) {
            bg.setImageBitmap(bitmap);
        }
        TextView name = findViewById(R.id.champion_name);
        name.setText(champion.getName());
        TextView title = findViewById(R.id.champion_title);
        title.setText(champion.getCapitalizedTitle());
        TextView blurb = findViewById(R.id.champion_blurb);
        blurb.setText(champion.getLore());
        supportStartPostponedEnterTransition();
    }
}

File: app/src/main/java/com/spiderbiggen/randomchampionselector/views/activities/ChampionActivity.kt
package com.spiderbiggen.randomchampionselector.views.activities

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import com.spiderbiggen.randomchampionselector.R
import com.spiderbiggen.randomchampionselector.data.cache.BitmapCache
import com.spiderbiggen.randomchampionselector.data.onMainThread
import com.spiderbiggen.randomchampionselector.domain.Champion
import kotlinx.android.synthetic.main.activity_champion.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.launch
import java.io.IOException

@ExperimentalCoroutinesApi
class ChampionActivity : AbstractActivity() {

    private var championKey = -1
    private var upOnBack: Boolean = false
    private var shouldFinalize = false

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        championKey = savedInstanceState?.getInt(CHAMPION_KEY) ?: championKey
        championKey = intent.getIntExtra(CHAMPION_KEY, championKey)
        upOnBack = intent.getBooleanExtra(UP_ON_BACK_KEY, true)
        setContentView(R.layout.activity_champion)
        setSupportActionBar(toolbar)
        supportPostponeEnterTransition()
        val actionBar = supportActionBar
        actionBar?.title = null
        actionBar?.setDisplayHomeAsUpEnabled(true)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putInt(CHAMPION_KEY, championKey)
        super.onSaveInstanceState(outState)
    }

    override fun onResume() {
        launch(Dispatchers.Default) {
            val champion = when {
                championKey < 0 -> database.findRandomChampion()
                else -> database.findChampion(championKey)
            }
            onMainThread { setChampion(champion) }
        }
        super.onResume()
    }

    override fun onStop() {
        if (shouldFinalize) {
            finish()
        }
        super.onStop()
    }


    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean =
        when (item?.itemId) {
            android.R.id.home -> {
                onBackPressed()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }

    override fun onBackPressed() {
        if (upOnBack) {
            val upIntent = parentActivityIntent
                ?: Intent(this, ListChampionsActivity::class.java)
            supportNavigateUpTo(upIntent)
            overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out)
        } else {
            supportFinishAfterTransition()
        }
    }

    fun openChampion(view: View) {
        openChampionActivity(view)
    }

    private fun setChampion(champion: Champion) {
        launch {
            try {
                val bitmap = BitmapCache.loadBitmap(champion)
                onMainThread { champion_splash.setImageBitmap(bitmap) }
            } catch (e: IOException) {
                onMainThread { loadImageFailure(e) }
            }
        }
        championKey = champion.key
        champion_name.text = champion.name
        champion_title.text = champion.capitalizedTitle
        champion_blurb.text = champion.lore
        supportStartPostponedEnterTransition()
    }

    private fun loadImageFailure(e: Throwable) {
        champion_splash.setImageBitmap(null)
        Log.e("ChampionActivity", "error ${e.message}", e)
    }

    companion object {
        const val CHAMPION_KEY = "CHAMPION_KEY"
        const val UP_ON_BACK_KEY = "UP_ON_BACK_KEY"
    }
}
